File: cpp/hot100/34.合并 K 个升序链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *merge(ListNode *l1,ListNode *l2){
        ListNode dummy(0);
        ListNode *cur=&dummy;
        while(l1&&l2){
            if(l1->val<l2->val){
                cur->next=l1;
                l1=l1->next;
            }
            else{
                cur->next=l2;
                l2=l2->next;
            }
            cur=cur->next;
        }
        cur->next=l1?l1:l2;
        return dummy.next;
    }
    ListNode *middle(vector<ListNode*> &lists,int l,int r){
        if(l==r) return lists[l];
        if(l>r) return nullptr;
        int mid=(l+r)/2;
        return merge(middle(lists,l,mid),middle(lists,mid+1,r));
    }

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        return middle(lists,0,lists.size()-1);
    }
};

File: python/hot100/34.合并 K 个升序链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergetwo(self,l1:Optional[ListNode],l2:Optional[ListNode])->Optional[ListNode]:
        cur=dummy=ListNode()
        while l1 and l2:
            if l1.val<l2.val:
                cur.next=l1
                l1=l1.next
            else:
                cur.next=l2
                l2=l2.next
            cur=cur.next
        cur.next=l1 if l1 else l2
        return dummy.next
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        length=len(lists)
        if length==0:
            return None
        if length==1:
            return lists[0]
        left=self.mergeKLists(lists[:length//2])
        right=self.mergeKLists(lists[length//2:])
        return self.mergetwo(left,right)